Fix trigger records in TRD trap simulator

bazinski opened the pull request:
- Fix the trigger records on receiving them in the trap simulator.
- boolean option --trd-fixtrigrecord  added to enable the fix if needed.
- if you request fix and it appears to not be needed, you are warned and nothing is done.
- fix double deletion error
- fix crash of trackletwriter
- removed object creation in message, and reverted to snapshotting at the end. Clearer to read as well.

sawenzel commented:
There seem to be commits in addition to "fix trigger records". Calibration is touched as well as new  macros added.  Consider mentioning this in the PR description / commit messages.

bazinski commented:
> There seem to be commits in addition to "fix trigger records". Calibration is touched as well as new macros added. Consider mentioning this in the PR description / commit messages.

@sawenzel those are carry overs from PR3649, this was branched off of trapsimlabels branch, once I rebase after the merge of PR3469 those will(should) all dissappear.

bazinski commented:
rebased and going back through the CI.
Back to 2 files changed.
@tdietel  small enough ? ;-)

jolopezl commented:
Do you really want to put a patch instead of waiting for the trigger records to be fixed from the TRD trigger?

bazinski commented:
> Do you really want to put a patch instead of waiting for the trigger records to be fixed from the TRD trigger?

Yip, as I understand it @martenole needs the timing of the tracklets. Its not on by default.

martenole commented on Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:
```diff
@@ -560,6 +585,7 @@ o2::framework::DataProcessorSpec getTRDTrapSimulatorSpec()
                              {"trd-printtrapconfig", VariantType::Bool, false, {"Name of the trap config from the CCDB"}},
                              {"trd-drawtracklets", VariantType::Int, 0, {"Bitpattern of input to TrapSimulator Draw method (be very careful) one file per track"}},
                              {"trd-printtracklets", VariantType::Int, 0, {"Bitpattern of input to TrapSimulator print method"}},
+                             {"trd-fixtrigrecord", VariantType::Bool, false, {"Fix trigger record alignment, temporary, hence false by default"}},
```
Should this not also be `trd-fixtriggerrecord`?

bazinski replied:
i can change it, i just wanted to type less ;-)

bazinski replied:
ah fixed.

martenole commented on Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:
```diff
@@ -301,6 +302,27 @@ bool digitindexcompare(unsigned int A, unsigned int B, const std::vector<o2::trd
     return 0;
 }
 
+void TRDDPLTrapSimulatorTask::fixTriggerRecords(std::vector<o2::trd::TriggerRecord>& trigRecord)
```
The initial TriggerRecord could in principle by zero and still they are shifted, right?

bazinski replied:
@martenole  I dont follow?

martenole replied:
Let's say the TriggerRecords are shifted by one. But by chance the first event actually contains zero tracklets. Then the fix would still be required but I would only get a warning.

bazinski replied:
the zero zero check is on the digits triggerrecord not the tracklet triggerrecord.

martenole replied:
Ok, I did not see that. Fine with me then

bazinski commented:
> I just tested this and I get a segfault from the TRDTrkltWrt class.

You using the same 10 event sample as before?
Rebuilding that branch now to check.

bazinski commented:
@martenole what simulation did you run this against, mine works.

o2-sim -n 10 -g pythia8 --skipModules ZDC 
o2-sim-digitizer-workflow -b
 o2-trd-trap-sim -b --trd-fixtriggerrecord

martenole commented:
Hi @bazinski , yes with exactly those three comands and this PR on top of dev I get a seg fault. It seems to be reproducible, but the error log is quite long. I don't want to post everything here.

[19209:TRAP]: [14:18:46][INFO] each trap position fired   : 105
[19209:TRAP]: [14:18:46][INFO] exiting the trap sim run method 
[INFO] Quitting
[19210:TRDTrkltWrt]: 
[19210:TRDTrkltWrt]:  *** Break *** segmentation violation
[INFO] QUIT_REQUESTED
[19209:TRAP]: [14:18:46][INFO] Received device shutdown request (signal 15).
[19209:TRAP]: [14:18:46][INFO] Waiting for graceful device shutdown. Hit Ctrl-C again to abort immediately.
[19211:internal-dpl-global-binary-file-sink]: [14:18:46][INFO] Received device shutdown request (signal 15).
[19211:internal-dpl-global-binary-file-sink]: [14:18:46][INFO] Waiting for graceful device shutdown. Hit Ctrl-C again to abort immediately.
[19211:internal-dpl-global-binary-file-sink]: [14:18:46][STATE] RUNNING ---> READY
[19211:internal-dpl-global-binary-file-sink]: [14:18:46][STATE] READY ---> RESETTING TASK
[19211:internal-dpl-global-binary-file-sink]: [14:18:46][STATE] RESETTING TASK ---> DEVICE READY
[19211:internal-dpl-global-binary-file-sink]: [14:18:46][STATE] DEVICE READY ---> RESETTING DEVICE
[19209:TRAP]: [14:18:46][STATE] RUNNING ---> READY
[19209:TRAP]: [14:18:46][STATE] READY ---> RESETTING TASK
[19209:TRAP]: [14:18:46][STATE] RESETTING TASK ---> DEVICE READY
[19209:TRAP]: [14:18:46][STATE] DEVICE READY ---> RESETTING DEVICE
[19207:internal-dpl-clock]: [14:18:46][INFO] Received device shutdown request (signal 15).
[19207:internal-dpl-clock]: [14:18:46][INFO] Waiting for graceful device shutdown. Hit Ctrl-C again to abort immediately.
[19208:TRDDIGITREADER]: [14:18:46][INFO] Received device shutdown request (signal 15).
[19208:TRDDIGITREADER]: [14:18:46][INFO] Waiting for graceful device shutdown. Hit Ctrl-C again to abort immediately.
[19210:TRDTrkltWrt]: [14:18:46][INFO] Received device shutdown request (signal 15).
[19210:TRDTrkltWrt]: [14:18:46][INFO] Waiting for graceful device shutdown. Hit Ctrl-C again to abort immediately.
[19207:internal-dpl-clock]: [14:18:46][STATE] RUNNING ---> READY
[19207:internal-dpl-clock]: [14:18:46][STATE] READY ---> RESETTING TASK
[19207:internal-dpl-clock]: [14:18:46][STATE] RESETTING TASK ---> DEVICE READY
[19207:internal-dpl-clock]: [14:18:46][STATE] DEVICE READY ---> RESETTING DEVICE
[19207:internal-dpl-clock]: [14:18:46][STATE] RESETTING DEVICE ---> IDLE
[19207:internal-dpl-clock]: [14:18:46][STATE] IDLE ---> EXITING
[19207:internal-dpl-clock]: [14:18:46][STATE] Exiting FairMQ state machine
[19207:internal-dpl-clock]: [14:18:46][INFO] Process 19207 is exiting.
[19211:internal-dpl-global-binary-file-sink]: [14:18:46][STATE] RESETTING DEVICE ---> IDLE
[19211:internal-dpl-global-binary-file-sink]: [14:18:46][STATE] IDLE ---> EXITING
[19211:internal-dpl-global-binary-file-sink]: [14:18:46][STATE] Exiting FairMQ state machine
[19211:internal-dpl-global-binary-file-sink]: [14:18:46][INFO] Process 19211 is exiting.
[19209:TRAP]: [14:18:46][STATE] RESETTING DEVICE ---> IDLE
[19209:TRAP]: [14:18:46][STATE] IDLE ---> EXITING
[19209:TRAP]: [14:18:46][STATE] Exiting FairMQ state machine
[19209:TRAP]: [14:18:46][INFO] Process 19209 is exiting.
[19209:TRAP]: double free or corruption (!prev)
[19210:TRDTrkltWrt]: 
[19210:TRDTrkltWrt]: 
[19210:TRDTrkltWrt]: 
[19210:TRDTrkltWrt]: ===========================================================
[19210:TRDTrkltWrt]: There was a crash.
[19210:TRDTrkltWrt]: This is the entire stack trace of all threads:
[19210:TRDTrkltWrt]: ===========================================================

bazinski commented:
ok will try a make clean and then repeat.

bazinski commented:
@martenole :-( ok I cant replicate, maybe chat on mattermost?

bazinski commented:
I cant replicate @martenole  problem.
I was getting one of : it varies each time you run it, but in the trap.

- free(): invalid pointer   --at the end
- double free or corruption (!prev)
- double free or corruption (out)

if I set MALLOC_CHECK_ to 2, it goes away. Still have not gotten the trackletwriter to crash though.
I updated alidist, reverted to dev and rebuilt O2 from scratch, now have a seg fault in TRDDigitReader, which has not been touched besides a log message since about October. Maybe its related?

sawenzel commented:
I confirm the segfaults seen by @martenole. Looking into it with @bazinski .

bazinski commented:
> I confirm the segfaults seen by @martenole. Looking into it with @bazinski .

I still cant replicate it :-( @bazinski is pulling what little hair I have left out ...
Thanks @sawenzel  for the backtraces. 

Do both of you build with --no-system ? Yip clutching a straws.

martenole commented:
@bazinski I simply build with `aliBuild build O2 --defaults o2`

bazinski commented:
@martenole can you check that now? Crash is fixed thanks to @sawenzel and triggerrecords themselves are fixed as well.

martenole commented:
@bazinski thanks a lot, this looks good now. For 10 pp events I get the following trigger records:

>Processing checkTRDTriggerRecords.C...
>There are 10 trigger records in the input with 730 tracklets in total
>Event 0: Occured at 104 us after SOR, contains 4 tracklets, index of first tracklet is 0
>Event 1: Occured at 113 us after SOR, contains 49 tracklets, index of first tracklet is 4
>Event 2: Occured at 124 us after SOR, contains 54 tracklets, index of first tracklet is 53
>Event 3: Occured at 178 us after SOR, contains 49 tracklets, index of first tracklet is 107
>Event 4: Occured at 230 us after SOR, contains 40 tracklets, index of first tracklet is 156
>Event 5: Occured at 230 us after SOR, contains 102 tracklets, index of first tracklet is 196
>Event 6: Occured at 231 us after SOR, contains 80 tracklets, index of first tracklet is 298
>Event 7: Occured at 272 us after SOR, contains 29 tracklets, index of first tracklet is 378
>Event 8: Occured at 286 us after SOR, contains 147 tracklets, index of first tracklet is 407
>Event 9: Occured at 296 us after SOR, contains 176 tracklets, index of first tracklet is 554
>Counted 730 tracklets referenced by the trigger records
>Done

The numbers add up now :+1:

tdietel commented:
> @tdietel: So it seems we can go on here?

Yes, it looks like everything is ok.
